XCUITests/FxScreenGraph.swift

```diff
@@ -435,8 +455,34 @@ func createScreenGraph(for test: XCTestCase, with app: XCUIApplication) -> Scree
         screenState.backAction = navigationControllerBackAction
     }
 
+    map.addScreenState(ChangePasscodeSettings) { screenState in
+        screenState.gesture(forAction: Action.ChangePasscode) { userState in
+            if let passcode = userState.passcode {
+                typePasscode(passcode)
+                typePasscode(userState.newPasscode)
+                typePasscode(userState.newPasscode)
+                userState.passcode = userState.newPasscode
+            }
+        }
+        screenState.gesture(forAction: Action.ConfirmPasscodeToChangePasscode) { userState in
+            if let passcode = userState.passcode {
+                typePasscode(passcode)
+            }
+        }
+        screenState.gesture(forAction: Action.ChangePasscodeTypeOnce) { userState in
+            typePasscode(userState.newPasscode)
+        }
+        screenState.backAction = navigationControllerBackAction
+    }
+
     map.addScreenState(LoginsSettings) { screenState in
         screenState.backAction = navigationControllerBackAction
+
+        screenState.gesture(forAction: Action.LoginPasscodeTypeOnce) { userState in
+                if let passcode = userState.passcode {
+                    typePasscode(passcode)
+                }
+        }
```

Comment:
I'm not sure I understand this gesture. It's conditional on `passcode != nil`.

I think I'd suggest a `LockedLoginsSettings` screen with a non-action gesture to `LoginsSettings`.

Actions probably should not be used for navigation on the whole, just changing the app state.

From `SettingsScreen`, link to `LockedLoginsSettings` and `LoginsSettings` but with conditions about `passcode != nil` and `passcode == nil` respectively.

Reply:
The idea behind creating it is that I need to go to LoginSettings when the passcode is set or not. And when it is set, I should be able to enter the correct passcode (this is easy using onEnter) or a wrong one (this is easy too with onEnter but only the first time), if being in that Enter Passcode screen (from Logins option) I did not know how to do it. 
So I tried a solution similar to SetPasscodeTypeOnce thinking that both were similar, but I may have misunderstood the SetPasscodeTypeOnce...

@jhugman would you have any suggestion about how to implement that? Thanks!
That is for example for test: testWrongPasscodeDisplaysAttemptsAndMaxError()

Reply:
Would it be fine if I define a enterpasscode funcion like the typepasscode but in navigator so that I can do:
navigator.goto(LockedLoginsSettings)
navigator.enterPasscode("111111")
and so select the different passcode that I want instead of doing it defining actions?

Reply:
I was thinking about two routes to get to LoginsSettings

 ```swift
        screenState.tap(table.cells["Logins"], to: LoginsSettings, if: "passcode == nil")
        screenState.tap(table.cells["Logins"], to: LockedLoginsSettings, if: "passcode != nil")
```

The `LoginsSettings` would be doing the actual work of managing logins (currently not a lot):

```swift
        map.addScreenState(LoginsSettings) { screenState in
                screenState.backAction = navigationControllerBackAction
        }
```

and the `LockedLoginsSettings` being the one that does the hard work of managing that lock screen
 
 ```swift
    map.addScreenState(LockedLoginsSettings) { screenState in
        // Add the actions about typing in the passcode incorrectly
        // so you can test error messages etc.
        screenState.gesture(forAction: Action.LoginPasscodeTypeOnce) { userState in
            typePasscode(userState.newPasscode)
        }

        // Add a gesture to get to the protected screen.
        screenState.gesture(to: LoginsSettings) { userState in
            typePasscode(userState.passcode)
        }

        // Add a back action.
        screenState.backAction = navigationControllerBackAction
    }
```

I expect the LoginsSettings screen isn’t the last place we’re going to want to protect with a passcode, so you might think about isolating this into a function
 
 ```swift
    func makeLocked(_ destinationScreen: String, typingAction: String) {
        // Add the actions about typing in the passcode incorrectly
        // so you can test error messages etc.
        screenState.gesture(forAction: typingAction) { userState in
            typePasscode(userState.newPasscode)
        }

        // Add a gesture to get to the protected screen.
        screenState.gesture(to: destinationScreen) { userState in
            typePasscode(userState.passcode)
        }
    }

    map.addScreenState(LockedLoginsSettings) { screenState in
        makeLocked(LoginsSettings, typingAction: Action.LoginsSettingsTypePasscode)

        // Add a back action.
        screenState.backAction = navigationControllerBackAction
    }
```

So now, the happy case works without worrying about passcodes:

```swift
navigator.goto(LoginsSettings) // passcode is not set
navigator.performAction(Action.SetPasscode)
navigator.goto(LoginsSettings) // passcode is set
```

But we don't lose the granularity when we want to worry about them.

```swift
navigator.performAction(Action.SetPasscode)
userState.newPasscode = "111112" // the wrong passcode
navigator.performAction(Action.LoginsSettingsTypePasscode) 
// … test the error message here.
```